doc/design/programming_solver.md

```diff
@@ -0,0 +1,141 @@
+# Mathematical Programming Solver with SQLFlow
+
+Mathematical programming (aka. [mathematical optimization](https://en.wikipedia.org/wiki/Mathematical_optimization)) is the selection of a best element (with regard to some criterion) from some set of available alternatives. Solving optimization problems is widely used in fields like ecnomics and finance, computer science, engineering and researching. In this design, we try to make SQLFlow be able to solve below branches of programming problems using SQL statements:
+
+- [Linear programming](https://en.wikipedia.org/wiki/Linear_programming)
+- [Quadratic programming](https://en.wikipedia.org/wiki/Quadratic_programming)
+- [Nonlinear programming](https://en.wikipedia.org/wiki/Nonlinear_programming)
+- [Integer programming](https://en.wikipedia.org/wiki/Integer_programming)
+- [Mixed-integer linear programming](http://macc.mcmaster.ca/maccfiles/chachuatnotes/07-MILP-I_handout.pdf)
+- [Mixed-integer quadratic programming](http://www.optimization-online.org/DB_FILE/2014/07/4446.pdf)
+- [Mixed-integer nonlinear programming](https://link.springer.com/book/10.1007/978-1-4614-1927-3)
+
+## Background
+
+To understand what's a mathematical programming problem, let's take a look at this example:
```

Comment:
I vaguely remember that I read this example somewhere. If this is not our original work, let us use an URL to the public post, instead of copying the content.  I am afraid that quotation without permission would cause some copyright issues.